tpl/images/images.go

```diff
@@ -117,3 +124,60 @@ func (ns *Namespace) Filter(args ...any) (images.ImageResource, error) {
 
 	return img.Filter(filtersv...)
 }
+
+type qrOptions struct {
+	Text      string // text to encode
+	Level     string // error correction level; one of low, medium, quartile, or high
+	TargetDir string // target directory relative to publishDir
+}
+
+var qrErrorCorrectionLevels = map[string]qr.Level{
+	"low":      qr.L,
+	"medium":   qr.M,
+	"quartile": qr.Q,
+	"high":     qr.H,
+}
+
+const qrDefaultErrorCorrectionLevel = "medium"
+
+// QR encodes text to a QR code using the given error correction level,
+// returning an image resource.
+func (ns *Namespace) QR(opts any) (images.ImageResource, error) {
+	qrOptions := qrOptions{
+		Level: qrDefaultErrorCorrectionLevel,
+	}
+
+	err := mapstructure.WeakDecode(opts, &qrOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if qrOptions.Text == "" {
+		return nil, errors.New("cannot encode an empty string")
+	}
+
+	level, ok := qrErrorCorrectionLevels[qrOptions.Level]
+	if !ok {
+		return nil, errors.New("error correction level must be one of low, medium, quartile, or high")
+	}
+
+	code, err := qr.Encode(qrOptions.Text, level)
+	if err != nil {
+		return nil, err
+	}
+
+	hash := hashing.XxHashFromStringHexEncoded(qrOptions.Text + qrOptions.Level)
```

Comment:
I suggest we create the hash from the input `opts`, e.g. `hash := hashing.HashString(opts)`. That way:

1. It will change whenever the user changes the options.
2. We don't have to worry about it when adding new options.

This is the pattern we use elsewhere (this becomes even more important if we want to cache to disk).

Reply:
Will do. The scaling thing works great, so I'm adding that option. Also, this is useful enough that I'd like to add an embedded shortcode. I know I've pushed back on adding more shortcodes in the past, but I think this is a non-opinionated addition. Please shout if you disagree.